src/data/list/basic.lean

```diff
@@ -572,6 +574,41 @@ begin
   { rw reverse_cons, exact H1 _ _ ih }
 end
 
+/-- Reassociates a list, revealing its last element. -/
+private def init_last_of_cons : ∀ (a : α) (l : list α), Σ' l' a', a :: l = l' ++ [a']
```

Comment:
I think a more useful version of this lemma (which shouldn't be private) is
```lean
lemma init_append_last {α} {l : list α} (h : l ≠ []) : init l ++ [last l h] = l
```
(the `l` in my lemma is your `a::l`)
With that lemma you should also be able to get your induction principle in a similar proof (or you could refactor it, showing that `lemma length_init {α} {l : list α} : length (init l) = length l - 1`)

Reply:
Ah, I was wracking my brain trying to get this to work with `init`/`last` – thanks for the suggestion!